### PR TITLE
Add Webpacker

### LINKED ISSUE #7774: Introduce Webpacker
This is required for StimulusReflex and possibly other parts of Product Uplift. It's also how all modern Rails apps handle assets...

It's a pretty serious change to the asset pipeline! It also requires some changes in the dev environment.

We could bring it in slowly and keep current assets as they are and introduce new assets in Webpacker, or we could go all in and move all of our Javascript and CSS into Webpacker all at once...

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

First part of #7774 

- Introduces and configures `Webpacker` for managing the asset pipeline
- Adds `foreman` and a `Procfile`. This enables multiple processes to be started at the same time in dev with a single command. Why? Using Webpacker in dev requires running a separate process called `webpack-dev-server`, and it's much nicer if it starts and stops automatically at the same time as the rails server (otherwise you have to manually start and stop it all the time in a separate terminal).

This step introduces Webpacker and means we can start using it. We can look at starting to move our current assets into Webpack but I think it's worth handling in a separate PR (or two) as it looks like it's going to be a bit crazy...

#### What should we test?

Nothing to test here really. Webpack is introduced but not actively used yet for serving our current assets.

#### Release notes

Introduced Webpacker for asset management.

Changelog Category: Technical changes

### CODE REVIEW

File: Procfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Foreman Procfile. Start all dev server processes with: `foreman start`
```
Comment (jibees) [inline Procfile:1]: 🎉 !

File: Procfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+rails:      bundle exec rails s -p 3000
+webpack:    ./bin/webpack-dev-server
+sidekiq:    bundle exec sidekiq -q mailers -q default
```
Comment (Matt-Yorkley) [inline Procfile:5]: Wondering what devs think about running Sidekiq by default when starting the rails server locally? 

Maybe it's not needed? I'd say 99% of the time I don't need the jobs queue in dev, which makes me think it doesn;t need to run by default.

On the other hand it doesn't really do any _harm_ to run it, and it means any delayed emails will actually get "delivered" (and open in the browser with letter-opener).

File: Procfile (hunk 1 above)
Comment (Matt-Yorkley) [reply Procfile:1]: Starting and stopping `webpack-dev-server` manually in a separate terminal every time you want to run `rails s`... nobody's got time for that, right?

Comment (jibees) [reply Procfile:1]: I did the same on a previous draft PR introducing webpacker!

File: Gemfile [hunk 3]
```diff
@@ -13,6 +13,7 @@ gem 'ransack', '2.4.1'
 gem 'responders'
 gem 'sass', '~> 3.4.0' # this restriction originates from foundation-rails's version
 gem 'sass-rails', '< 5.1.0' # this restriction originates from the compass-rails's version
+gem 'webpacker', '~> 5'
```
Comment (luisramos0) [inline Gemfile:16]: why ~> 5? Can we leave it open? we will have dependabot making upgrades and we will see if v6 will work or not.

File: postcss.config.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    require('postcss-import'),
+    require('postcss-flexbugs-fixes'),
+    require('postcss-preset-env')({
+      autoprefixer: {
+        flexbox: 'no-2009'
+      },
+      stage: 3
+    })
+  ]
+}
```
Comment (luisramos0) [inline postcss.config.js:12]: do we need this? we use scss right?

File: babel.config.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      ],
+      [
+        '@babel/plugin-transform-regenerator',
+        {
+          async: false
+        }
+      ]
+    ].filter(Boolean)
+  }
+}
```
Comment (luisramos0) [inline babel.config.js:70]: can we delete this file? we dont use babel right?

**Review by luisramos0 — APPROVED** [review]
It's a goodstart!

File: Gemfile (hunk 3 above)
Comment (Matt-Yorkley) [reply Gemfile:16]: I think there's various breaking changes in `6.x`, it won't be a simple case of bumping the version.

File: babel.config.js (hunk 5 above)
Comment (Matt-Yorkley) [reply babel.config.js:70]: > can we delete this file? we dont use babel right?

I have no idea! :sweat_smile:  These files are added by webpacker when it's installed. I'm not sure if webpacker itself uses babel in some way?

File: Gemfile (hunk 3 above)
Comment (luisramos0) [reply Gemfile:16]: ok :+1:

File: postcss.config.js (hunk 4 above)
Comment (Matt-Yorkley) [reply postcss.config.js:12]: > we use scss right?

Yes, but I think PostCSS is part of the toolchain by default and this file is used to configure it. So for example you can add post-processing functions on the compiled output (even if you're using SCSS)..?

File: babel.config.js (hunk 5 above)
Comment (jibees) [reply babel.config.js:70]: Webpacker can use babel (to transpile ES6/7). Then, I think it's the default conf.

File: postcss.config.js (hunk 4 above)
Comment (jibees) [reply postcss.config.js:12]: Like a minimifier I guess.

**Review by jibees — APPROVED** [review]
👌

File: Procfile (hunk 2 above)
Comment (andrewpbrett) [reply Procfile:5]: I've found it helpful with the current issue I'm working on, so I've gotten in the habit of starting it any time I start fresh even if I don't know for sure that I'll need it. It's handy to know when emails get sent and might help at some point to notice a change in behavior if they don't go out. I'd vote to have it start by default.

**Comment by andrewpbrett** [comment]
This looks like it's ready to merge, apart from maybe a merge resolution (I'm guessing we rebase and do a fresh `yarn install` to resolve something like this?).

**Comment by Matt-Yorkley** [comment]
> I'm guessing we rebase and do a fresh yarn install to resolve something like this?

Yep, `yarn` is really good at resolving merge conflicts automatically :tada:

Screenshot from 2021-06-19 17-53-39

**Comment by filipefurtad0** [comment]
Hey team,

I think this PR might be breaking deployments, see logs on both staging-AU and -FR, see https://github.com/openfoodfoundation/openfoodnetwork/pull/7281#issuecomment-865780320.


`
Module build failed (from ./node_modules/babel-loader/lib/index.js):\nError: Cannot find module '@babel/plugin-transform-destructuring'\nRequire stack:\n- /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2021-06-22-072836/node_modules/@babel/core/lib/config/files/plugins.js\n- /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2021-06-22-072836/node_modules/@babel/core/lib/config/files/index.js`

and 

`
["RAILS_ENV=staging environment is not defined in config/webpacker.yml, falling back to production environment", "Compiling...", 

"Compilation failed:", "Hash: e78cde7e6cec8054a849", "Version: webpack 4.46.0", "Time: 4842ms", "Built at: 06/22/2021 7:39:34 AM", " 2 assets", "Entrypoint application = js/application-9ee1c398c69079370bd8.js js/application-9ee1c398c69079370bd8.js.map", "[0] ./app/webpacker/packs/application.js 5.57 KiB {0} [built] [failed] [1 error]", "", "ERROR in ./app/webpacker/packs/application.js"`

I think this was merged after the release was drafted, so we should be clear on that side.

File: Procfile (hunk 2 above)
Comment (sauloperez) [reply Procfile:5]: To me, this looks almost like an exact duplicate of https://github.com/openfoodfoundation/openfoodnetwork/blob/master/docker-compose.yml. If I'm the only one using it, we better drop support for it. No point in maintaining something soon no one will use.

**Comment by Matt-Yorkley** [comment]
Should be fixed now :+1:

File: Gemfile [hunk 6]
```diff
@@ -162,6 +163,7 @@ end
 
 group :development do
   gem 'debugger-linecache'
+  gem 'foreman'
```
Comment (mkllnk) [inline Gemfile:166]: Just noting that the author of the gem recommends against this:

https://github.com/ddollar/foreman/wiki/Don't-Bundle-Foreman

Comment (Matt-Yorkley) [reply Gemfile:166]: I did notice that, but if we don't add it in the dev group then devs just have to manually install the gem. I don't see any benefit there?

Looking through the history of the repo I think the gem had a lot more dependencies in the past, and that was the reason for keeping it out of the Gemfile, but most of those dependencies have been stripped out over time and it's quite minimal now.

**Pull request merged**